

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>References &mdash; aPhyloGeo  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=5929fcd5"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Alignment Dependencies Installation" href="alignment_dependencies.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            aPhyloGeo
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Available Analyses</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="sequence-similarity-algorithms.html">Sequence Similarity Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="metrics.html">Metrics</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Api</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="api/genetic_trees.html">aphylogeo.genetic_trees</a></li>
<li class="toctree-l1"><a class="reference internal" href="api/alignement.html">aphylogeo.alignement</a></li>
<li class="toctree-l1"><a class="reference internal" href="api/multiProcessor.html">aphylogeo.multiProcessor</a></li>
<li class="toctree-l1"><a class="reference internal" href="api/params.html">aphylogeo.params</a></li>
<li class="toctree-l1"><a class="reference internal" href="api/utils.html">aphylogeo.utils</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Misc</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="user-preferences.html">Editing User Preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="alignment_dependencies.html">Alignment Dependencies Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">References</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#state-of-the-field-advancements-in-genomic-analysis">State of the Field - Advancements in Genomic Analysis</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">aPhyloGeo</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">References</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/tahiri-lab/aPhyloGeo/blob/main/docs/source/references.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="references">
<span id="id1"></span><h1>References<a class="headerlink" href="#references" title="Link to this heading"></a></h1>
<section id="state-of-the-field-advancements-in-genomic-analysis">
<h2>State of the Field - Advancements in Genomic Analysis<a class="headerlink" href="#state-of-the-field-advancements-in-genomic-analysis" title="Link to this heading"></a></h2>
<p>The field of genomic analysis has progressed significantly in recent years, notably in the creation of tools and algorithms to explore the intricate relationship between genetic variation and environmental factors. Our algorithm for identifying sub-sequences within genes [#nadia_tahiri-proc-scipy-2022], and its subsequent application to SARS-CoV-2 data in 2023 [#nadia_tahiri-proc-scipy-2023], enhance comprehension of the genetic underpinnings of adaptation across various species and environments.</p>
<p>In the broader field of phylogeography, substantial methodological advancements have also occurred. Several Python packages provide functionalities pertinent to phylogeographic analysis, but often in a fragmented way. <a class="reference external" href="https://pypi.org/project/biopython">Biopython</a>, a cornerstone in bioinformatics, excels at handling genetic sequences and basic phylogenetic tasks, yet falls short in integrating environmental data. <a class="reference external" href="https://pypi.org/project/DendroPy">DendroPy</a>, a robust library for phylogenetic trees, aids in visualizing phylogeographic patterns but requires additional tools for comprehensive analysis. While <a class="reference external" href="https://pypi.org/project/scipy/">SciPy</a>’s statistical utilities could be harnessed for custom analyses, its complexity demands a strong background in statistical programming. <a class="reference external" href="https://pypi.org/project/geopandas/">GeoPandas</a>, adept at handling geospatial data, is useful for mapping genetic or environmental distributions, but lacks seamless integration with genetic data analysis tools. In summary, while powerful individual tools exist, a comprehensive and user-friendly Python package specifically designed for phylogeographic analysis remains a gap to be filled.</p>
<p>Statistical approaches, including generalized linear models (GLMs) and mixed models, are increasingly used to investigate the relationship between genetic variation and environmental variables. These methods enable researchers to quantify the relative influence of various factors, such as climate, geography, and demography, on observed patterns of genetic diversity.</p>
<p>The continuous refinement of these tools and methodologies, coupled with the growing availability of high-throughput sequencing technologies and environmental data, has opened up exciting new research avenues in evolutionary biology, ecology, and conservation. <em>aPhyloGeo</em> builds upon these advancements, providing a unified platform for integrating genetic and climatic data to address a wide array of phylogeographic questions. By bridging the gap between genomics and environmental science, <em>aPhyloGeo</em> aims to contribute to a more comprehensive understanding of the forces shaping biodiversity in a changing world.</p>
<ol class="arabic simple">
<li><p><strong>Calculation of distance between phylogenetic tree: `Least Square metric`</strong></p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1706274/">Cavalli-Sforza, L. L., &amp; Edwards, A. W. (1967). Phylogenetic analysis. Models and estimation procedures. American journal of human genetics, 19(3 Pt 1), 233.</a></p></li>
<li><p><a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/11975348/">Felsenstein, J. (1997). An alternating least squares approach to inferring phylogenies from pairwise distances. Systematic biology, 46(1), 101-111.</a></p></li>
<li><p><a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/15059836/">Makarenkov, V., &amp; Lapointe, F. J. (2004). A weighted least-squares approach for inferring phylogenies from incomplete distance matrices. Bioinformatics, 20(13), 2113-2121.</a></p></li>
</ul>
</li>
<li><p><strong>Calculation of distance between phylogenetic tree: `Robinson-Foulds metric`</strong></p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.sciencedirect.com/science/article/abs/pii/0025556481900432?via%3Dihub">Robinson, D.F. and Foulds, L.R., 1981. Comparison of phylogenetic trees. Mathematical biosciences, 53(1-2), pp.131-147.</a></p></li>
</ul>
</li>
<li><p><strong>Dataset full description: `Analysis of genetic and climatic data of SARS-CoV-2`</strong></p>
<ul class="simple">
<li><p><a class="reference external" href="https://conference.scipy.org/proceedings/scipy2022/nadia_tahiri.html">Koshkarov, A., Li, W., Luu, M. L., &amp; Tahiri, N. (2022). Phylogeography: Analysis of genetic and climatic data of SARS-CoV-2.</a></p></li>
<li><p><a class="reference external" href="https://conference.scipy.org/proceedings/scipy2023/nadia_tahiri.html">Li, W., &amp; Tahiri, N. (2023). aPhyloGeo-Covid: A Web Interface for Reproducible Phylogeographic Analysis of SARS-CoV-2 Variation using Neo4j and Snakemake.</a></p></li>
</ul>
</li>
<li><p><strong>Muscle5</strong>:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://doi.org/10.1101/2021.06.20.449169">Edgar, RC (2021). MUSCLE v5 enables improved estimates of phylogenetic tree confidence by ensemble bootstrapping, bioRxiv 2021.06.20.449169.</a></p></li>
</ul>
</li>
<li><p><strong>Fastree</strong>:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://doi.org/10.1371/journal.pone.0009490">Price, MN, Dehal, PS, and Arkin, AP (2010). FastTree 2 – Approximately Maximum-Likelihood Trees for Large Alignments, PLoS ONE 5(3): e9490.</a></p></li>
</ul>
</li>
<li><p><strong>ClustalW</strong>:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://doi.org/10.1093/bioinformatics/btm404">Larkin, MA, Blackshields, G, Brown, NP, Chenna, R, McGettigan, PA, McWilliam, H, Valentin, F, Wallace, IM, Wilm, A, Lopez, R, Thompson, JD, Gibson, TJ, and Higgins, DG (2007). Clustal W and Clustal X version 2.0, Bioinformatics 23(21): 2947-2948.</a></p></li>
</ul>
</li>
<li><p><strong>Mafft</strong>:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://doi.org/10.1093/bib/bbx108">Katoh, K, Rozewicki, J, and Yamada, KD (2019). MAFFT online service: multiple sequence alignment, interactive sequence choice and visualization, Briefings in Bioinformatics 20(4): 1160-1166.</a></p></li>
</ul>
</li>
</ol>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="alignment_dependencies.html" class="btn btn-neutral float-left" title="Alignment Dependencies Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Tahiri Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>